python/essex-config/essex_config/config.py

```diff
@@ -0,0 +1,297 @@
+"""Main configuration module for the essex_config package."""
+
+from functools import cache
+from typing import Any, ClassVar, TypeVar, cast
+
+from pydantic import BaseModel
+from pydantic_core import PydanticUndefined
+
+from essex_config.configuration_field import ConfigurationField, FieldVisibility
+from essex_config.sources import EnvSource, Source
+
+DEFAULT_SOURCE_LIST: list[Source] = [EnvSource()]
+
+T = TypeVar("T", bound="Config")
+
+
+class Config(BaseModel):
+    """Base class for configuration.
+
+    Classes that inherit from this class will have their fields populated with values from the sources defined in the sources attribute.
+    The sources class attribute is a list of `Source` objects that will be used to get the configuration values. The default value is a list with the `EnvSource` object.
+    The fields of the class should be annotated with the `Annotated` type from the `typing` module. The first argument of the `Annotated` type should be the type of the field and the second argument should be an instance of the `ConfigurationField` class.
+
+    Examples of usage:
+
+    ```python
+    from essex_config.config import (
+        Config,
+    )
+    from essex_config.configuration_field import (
+        ConfigurationField,
+        FieldVisibility,
+    )
+    from essex_config.sources.source import (
+        Source,
+    )
+
+    # Define a class that inherits Config class
+    # In this case this class will use the default sources, meaning it will fetch the values from the environment variables
+    # In this case the environment was set to:
+    # value1=value1
+    # value2=2
+    # value3=3.0
+    # value4=True
+
+
+    class BasicConfig(
+        Config
+    ):
+        value1: Annotated[
+            str,
+            ConfigurationField(),
+        ]
+        value2: Annotated[
+            int,
+            ConfigurationField(),
+        ]
+        value3: Annotated[
+            Any,
+            ConfigurationField(),
+        ]
+        value4: Annotated[
+            bool,
+            ConfigurationField(),
+        ]
+
+
+    config = BasicConfig.get_config()
+
+    assert (
+        config.value1
```

Comment:
If by any chance, we end up having a `some_list: list[AnotherConfig]` as an attribute. How would something like this behave: `config_a.sub_config_a.some_list[3].sub_config_b.some_value` ?

I am guessing it just works because we are talking about a base model from pydantic right?

Reply:
This is an excellent question, I think it will ignore this field as the code is right now, should we allow for this to happen? what would be the use case for this?